### ISSUE TITLE
nn.Transformer tutorial uses nn.TransformerEncoder only

### ISSUE THREAD

**Issue by vainaijr** [pytorch/examples#671]
hello,
when I search for nn.Transformer use example, I find example which uses nn.TransformerEncoder, is there example use of nn.Transformer?

**Comment by zhangguanheng66** [comment]
We didn't use `nn.Transformer` in the example. However, if you want, you could apply it in the same problem. Let's say you have the training data as "train the model with transformer". Then, feed the src sequence "train the model with transformer" to the encoder side and "<BOL> train the model with" to the decoder side. Compare the output of transformer with "train the model with transformer".

**Comment by vainaijr** [comment]
I use nn.Transformer here, is it correct way?
https://discuss.pytorch.org/t/is-this-a-correct-use-of-nn-transformer/62542

**Comment by zhangguanheng66** [comment]
I took a look at your code. It seems like a right way. However, depending on your problem, you may need `mask` to shield the inputs. Could you briefly describe your problem here?

**Comment by vainaijr** [comment]
I check nn.Transformer for predicting next item, on a toy dataset, that is for 0, 1, 0, 1, 0, 1, I get two item sequence 0, 1, and for 1, 0, 1, 0, 1, 0, I get two item sequence 1, 0.

The softmax probabilities represent whether neural network predicts 1 or 0 for given input sequence.

But is it correct way if I already specify the output in dataset, and should I set target embedding requires_grad to False?

Or am I making some mistake here, and Transformer is not applicable for such a problem.

**Comment by zhangguanheng66** [comment]
OK, I see. You want to extract the pattern from a sequence. IMO, during training, you could pass two constants to the decoder and the input sequence to the encoder. Then, feed the output of decoder to `CrossEntropyLoss` together with the target sequence.

**Comment by vainaijr** [comment]
But if I use nn.Transformer directly, not separating into encoder and decoder, then I have to pass input and target both to nn.Transformer?
And in crossentropy, specify target once again?

**Comment by hirabarahyt** [comment]
> We didn't use `nn.Transformer` in the example. However, if you want, you could apply it in the same problem. Let's say you have the training data as "train the model with transformer". Then, feed the src sequence "train the model with transformer" to the encoder side and " train the model with" to the decoder side. Compare the output of transformer with "train the model with transformer".

I am not sure that feed "train the model with" to the decoder side makes the length of transformer's output is 4. So maybe compare output of transformer with "the model with transformer" ?

**Comment by gemfield** [comment]
There is a nn.Transformer example: https://github.com/deepvac/transformer, including the end-to-end training code.
